api/src/main/java/org/pmiops/workbench/ras/RasLinkService.java

```diff
@@ -0,0 +1,147 @@
+package org.pmiops.workbench.ras;
+
+import static org.pmiops.workbench.ras.OpenIdConnectClient.decodedJwt;
+import static org.pmiops.workbench.ras.RasLinkConstants.ACR_CLAIM;
+import static org.pmiops.workbench.ras.RasLinkConstants.EMAIl_FIELD_NAME;
+import static org.pmiops.workbench.ras.RasLinkConstants.Id_TOKEN_FIELD_NAME;
+import static org.pmiops.workbench.ras.RasLinkConstants.LOGIN_GOV_IDENTIFIER_LOWER_CASE;
+import static org.pmiops.workbench.ras.RasLinkConstants.PREFERRED_USERNAME_FIELD_NAME;
+import static org.pmiops.workbench.ras.RasLinkConstants.RAS_AUTH_CODE_SCOPES;
+import static org.pmiops.workbench.ras.RasOidcClientConfig.RAS_OIDC_CLIENT;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.NullNode;
+import com.google.api.client.auth.oauth2.TokenResponse;
+import java.io.IOException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.pmiops.workbench.db.dao.UserService;
+import org.pmiops.workbench.db.model.DbUser;
+import org.pmiops.workbench.exceptions.ForbiddenException;
+import org.pmiops.workbench.exceptions.ServerErrorException;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Service;
+
+/**
+ * Service handles link login.gov account with All of Us account. It finishes OAuth dance with RAS
+ * then validate users use their login.gov account with IAL2 enabled.
+ *
+ * <p>Key steps: Step1: Finish OAuth and get {@link TokenResponse}. A sample response:
+ *
+ * <pre>{@code
+ * "access_token":"JWT token"
+ * "token_type":"Bearer",
+ * "expires_in":1800,
+ * "refresh_token":"refresh token",
+ * "scope":"openid profile email ga4gh_passport_v1",
+ * "sub":"subId",
+ * "id_token":"JWT token"
+ * "id_token_type":"urn:ietf:params:oauth:grant-type:jwt-bearer"
+ * }</pre>
+ *
+ * Step2: Decode JWT id_token to extract IAL status. Decoded id_token:
+ *
+ * <pre>{@code
+ * "sub": "subId",
+ * "aud": "client id",
+ * "c_hash": "",
+ * "acr": "https://stsstg.nih.gov/assurance/ial/1 https://stsstg.nih.gov/assurance/aal/1",
+ * "azp": "",
+ * "auth_time": 1615910504,
+ * "iss": "https://stsstg.nih.gov",
+ * "exp": 1615996939,
+ * "iat": 1615910539
+ * }</pre>
+ *
+ * The acr claim contains user IAL status, the number after /assurance/ial/
+ *
+ * <p>Step3: User access_token to pull RAS user info. A sample userInfo in Json format:
+ *
+ * <pre>{@code
+ * "sub":"subId",
+ * "name" : "",
+ * "preferred_username":"user1@Login.Gov",
+ * "userId":"user1"
+ * "email":"foo@gmail.com"
```

Comment:
Is this just a random contact email value that we don't care about?

Reply:
that is the email we use to register and login in login.gov(serve as username). Based on my experience, that seems to be immutable, and unique per user. (you need to delete that account if you want to re-use that emai), and that is the only meaningful username we can get from login.gov. So I use that as login.gov username.
Another option is use user id, but I think that is just a randomUUID, that is not as useful as email.